### ISSUE TITLE
Features dependencies improvments

### ISSUE THREAD

**Issue by jtkech** [OrchardCMS/OrchardCore#8955]
@sebastienros 

Because of some recent issues e.g. #8942 and #8785 i started to work / think about

- If we upgrade an already enabled feature whose coded dependencies have changed, we could think about auto enableing the features included in the dependencies that are not yet enabled.

- If F1 requires F2 to add a functionality, **about ordering** this functionality would need to be a pseudo feature that depends as usual on the main one but that would also depends on F2. It could be a feature like the others, but auto enabled / disabled based on the state of its dependencies.

**I can work on these kind of things, but only if you think it is important and worth to do it ;)**

---

@deanmarcussen For infos, recently in a migration step we enable a feature in a deffered task, this task will call `FireApplyChangesIfNeeded` that also uses a deferred task, but a deffered task can't register a deffered task. Not so important even it will not update the ShellState (not the ShellDescriptor) because we don't use it elsewhere. So just to say that maybe it is not the best way (my bad), and could be solved more globally as suggested above.

**Comment by sebastienros** [comment]
> If we upgrade an already enabled feature whose coded dependencies have changed, we could think about auto enabling the features included in the dependencies that are not yet enabled.

Why "if we upgrade". Can't it be done at startup? Just check that all dependent feature are enabled, and enable them. This is what would have happened if we had enabled the feature in the first place, from the UI for instance.

Either that or we don't enable the feature since it can't work without its dependencies, but that could block the system.

**Comment by jtkech** [comment]
Yes, this is what we do currently in a migration step

But not so simple, when enableing through the UI, from memory

- Enable the feature + update shell descriptor + release the shell
- Redirect
- A defferred task update the ShellState (not ShellDescriptor) in the db (but for now we don't use it elsewhere)
- Next request rebuilds the shell (Startups ConfigureServices are called)
    - Activates it (where a migration may happen and may need the dependency)
    - Then rebuilds the shell pipeline (Startups Configure are called)

I also talked about some ordering executions related to features using the `RequireFeatures` attribute

**Comment by jtkech** [comment]
I began to have some ideas but need more thinkings, maybe just some helpers to do things but only when there are needed, so for now i close it and will open a new issue / PR when ready to rephrase it.

**Comment by deanmarcussen** [comment]
> If we upgrade an already enabled feature whose coded dependencies have changed, we could think about auto enableing the features included in the dependencies that are not yet enabled.

Can we not just have a hash of current features vs discovered features (i.e. a new feature was added / and or a dependency hash changed), and then invoke the renable flow on startup? 

Sounds super simple when I say it, probably is not?

>  recently in a migration step we enable a feature in a deffered task, 

ok, so I was able to do this without a deffered task, just in the same shell scope, so maybe that is better?  I will look at shortcodes today, so will see what actually is happening there (name change, or feature missing, not sure yet), and we can reconvene on the layers deferred task if necesary.

**Comment by jtkech** [comment]
@deanmarcussen 

There is no problem to check missing enabled features based on the actual dependencies, the "problem" is more to know when doing things, particularly when there are related migrations.

> ok, so I was able to do this without a deffered task, just in the same shell scope

Yes i remember, but nothing urgent, as said there is a shell state that is not updated but currently it is not used elsewhere, maybe we needed this shell state when we had different module loading strategy, e.g. with dynamic compilation.

But apparently you still needed to do it at the end of the scope. Would it be okay in your tests if the `Rules` feature was enabled before? Or does it really need to be at the end? One problem is when we enable a feature we also need to activate it (e.g. will run its own migrations) this through a shell activation, but being in a migration step means that we are activating the shell. This is this kind of things that need to be analysed and tried, ATM the first idea that comes to me is to enable it before all the migrations, maybe at the shell host level, not the shellscope one.

Interesting, in the layer migration you inject the `_ruleMigrator` that is registered by a feature that is not yet enabled ;) It works because the actual dependencies in the shell blueprint already include the `Rules` feature and then its related types including the related `IStartup` that has run. But still important to enable the feature for other components, e.g. to run the related migrations ;)

**What is important to know is if we can use a database table during migrations that is created before in the same migrations scope**, seems okay as you said to me for an index table that will be still populated, so i assume it is the same for any table, i would just have to try it ;)

If it's okay, the only main problem would be the order of executions, e.g. with the `RequireFeatures`. Otherwise we could think about using a shell scope for each migration step, as we already do for regular recipe steps, but would need to be tried to see if it is okay.

Sorry for all these unordered thinkings, maybe in summary we jut need, before running migrations, that all needed features are enabled, even if we need to recompose the shell, as we do at the end of a setup, then for the `RequireFeatures` think about some tweaks about the dependency ordering, a pseudo feature (a conditional functionality of an existing feature) requiring another feature would need to implicitly depends on it.
